mautic/mautic
# Adding an option to hide themes

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ❌ 
| New feature/enhancement? (use the a.x branch)      | ✅ 
| Deprecations?                          | ❌ 
| BC breaks? (use the c.x branch)        | ✅ 
| Automated tests included?              | ✅  
| Related user documentation PR URL      | https://github.com/mautic/user-documentation/pull/303
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | /



#### Description:


Give user option to hide/unhide 'default' themes in the themes list page
![Screenshot 2024-05-06 at 12 15 31](https://github.com/mautic/mautic/assets/1235442/496ff2c5-fb1d-4f40-9a62-419e16acb98f)
![Screenshot 2024-05-06 at 12 15 40](https://github.com/mautic/mautic/assets/1235442/f88a4542-cac6-444c-ac04-c901fb33105a)
![Screenshot 2024-05-06 at 12 15 55](https://github.com/mautic/mautic/assets/1235442/a3a5d204-bb90-4e13-adb2-8e5f1d4daca1)

 Do no show these hidden themes in the Email/Page theme list
Update existing themes using migration

#### Steps to test this PR:


MAUT-10202:
1. Go to themes, check if 'Hide' option is visible for the default themes only
2. On click of 'Hide' option in the dropdown, confirmation popup appears.
3. On confirmation, theme is moved to the bottom of the theme list and is disabled (shown in grey font) and theme is hidden from email/page theme list and also in the 
5. Hidden themes will have 'Unhide' menu option, on clicking confirmation popup opens, on confirming theme will unhide and will be visible in the previous order and also visible in the email/page theme list
6. We have added a migration to rename deleted.txt with hidden-themes.txt and remove custom theme names from the file which are present in the instance's `public_html/themes` directory.

MAUT-10140:
1. There is no option to delete default themes anymore (this is now changed to Hide/Unhide)
2. Custom themes can be deleted, these won't be added to deleted.txt (deleted.txt has been renamed to hidden-themes.txt)
3. Saving custom themes with already existing names will generate error for e.g. Template name `abc` already exists.
4. Since, saving a theme is prefixed with 'beefree-' by default, if we try to save a befree theme without the 'beefree-' keyword, will generate error for e.g. If we try to save them with name `fall-in-love-with-this-seasons-collection` will generate - beefree-fall-in-love-with-this-seasons-collection is the default theme and therefore cannot be overwritten.
